test/e2e/storage/csi_mock/csi_volume_expansion.go

```diff
@@ -209,4 +249,248 @@ var _ = utils.SIGDescribe("CSI Mock volume expansion", func() {
+				pvcRequestSize:          "9Gi",
+				allocatedResource:       "9Gi",
+				expectedExpansionStatus: expansionFailedOnNode,
+				resizeStatus:            v1.PersistentVolumeClaimNodeExpansionFailed,
+				recoverySize:            resource.MustParse("5Gi"),
+			},
+		}
+
+		for _, t := range tests {
+			test := t
+			ginkgo.It(test.name, func() {
+				var err error
+				params := testParameters{enableResizing: true, enableNodeExpansion: true, enableRecoverExpansionFailure: true}
+				if test.disableAttach {
+					params.disableAttach = true
+					params.registerDriver = true
+				}
+
+				if test.expectedExpansionStatus != expansionSuccess {
+					params.hooks = createExpansionHook(test.expectedExpansionStatus)
+				}
+
+				m.init(params)
+
+				defer m.cleanup()
+
+				sc, pvc, pod := m.createPod(pvcReference)
+				gomega.Expect(pod).NotTo(gomega.BeNil(), "while creating pod for resizing")
+
+				if !*sc.AllowVolumeExpansion {
+					framework.Fail("failed creating sc with allowed expansion")
+				}
+
+				err = e2epod.WaitForPodNameRunningInNamespace(m.cs, pod.Name, pod.Namespace)
+				framework.ExpectNoError(err, "Failed to start pod1: %v", err)
+
+				ginkgo.By("Expanding current pvc")
+				newSize := resource.MustParse(test.pvcRequestSize)
+				newPVC, err := testsuites.ExpandPVCSize(pvc, newSize, m.cs)
+				framework.ExpectNoError(err, "While updating pvc for more size")
+				pvc = newPVC
+				gomega.Expect(pvc).NotTo(gomega.BeNil())
+
+				pvcSize := pvc.Spec.Resources.Requests[v1.ResourceStorage]
+				if pvcSize.Cmp(newSize) != 0 {
+					framework.Failf("error updating pvc size %q", pvc.Name)
+				}
+
+				switch test.expectedExpansionStatus {
+				case expansionSuccess:
+					validateExpansionSuccess(pvc, m, test, test.allocatedResource)
+				case expansionFailedOnController:
+					checkControllerExpansionFailureRecovery(pvc, m, test)
+				case expansionFailedOnNode:
+					checkNodeExpansionFailure(pvc, m, test, pod)
+				}
+			})
+		}
+
+	})
 })
+
+func checkControllerExpansionFailureRecovery(pvc *v1.PersistentVolumeClaim, m *mockDriverSetup, test recoveryTest) {
+	var err error
+	ginkgo.By("Waiting for resizer to set allocated resource")
+	err = waitOnAllocatedResource(pvc, m, test.allocatedResource)
+	framework.ExpectNoError(err, "While waiting for allocated resource to be updated")
+
+	ginkgo.By("Waiting for resizer to set resize status")
+	err = waitForResizeStatus(pvc, m.cs, test.resizeStatus)
+	framework.ExpectNoError(err, "While waiting for resize status to be set")
+
+	ginkgo.By("Recover pvc size")
+	newPVC, err := testsuites.ExpandPVCSize(pvc, test.recoverySize, m.cs)
+	framework.ExpectNoError(err, "While updating pvc for more size")
+	pvc = newPVC
+	gomega.Expect(pvc).NotTo(gomega.BeNil())
+
+	pvcSize := pvc.Spec.Resources.Requests[v1.ResourceStorage]
+	if pvcSize.Cmp(test.recoverySize) != 0 {
+		framework.Failf("error updating pvc size %q", pvc.Name)
+	}
```

Comment:
This is the same as in `checkNodeExpansionFailure` below. Would it make sense to add `expectedRecoveryFailure` to the test struct and merge `checkNodeExpansionFailure` and `checkControllerExpansionFailureRecovery` together?

Reply:
fixed.